Allow integer indexing

thorio opened the pull request:
Closes #92.

This PR contains a few related changes for various issues:
- Indexing arrays in `Value::find()` (#92)
- Providing partial array values from Env (#113)
- Necessitated by the Env change, new strategies zipjoin/zipmerge, which allow partially overwriting arrays and values inside arrays.

thorio committed:
Make 'Value::find()' able to traverse arrays.

thorio committed:
Make 'nest()' consider array indices.

thorio committed:
Add 'Figment::zipjoin()' and 'Figment::zipmerge()'.

amircodota commented:
@thorio ran a quick test with [my use case](https://github.com/SergioBenitez/Figment/issues/113)

Works like a charm for both top level arrays and also arrays in sub-keys :crown:

thorio commented:
@SergioBenitez could you take a look at this?

SergioBenitez commented on src/value/value.rs:
```diff
@@ -394,6 +426,35 @@ impl Value {
+    pub(crate) fn index(self, key: &str) -> Option<Value> {
+        fn try_remove<T>(mut vec: Vec<T>, key: &str) -> Option<T> {
+            let index = key.parse().ok()?;
+            vec.get(index)?;
+
+            Some(vec.swap_remove(index))
```
This hides a bit of the logic. I'd prefer something like:

```suggestion
            match vec.get(index) {
                Some(_) => Some(vec.swap_remove(index)),
                None => None,
            }
```

Unfortunately the `take()` methods are unstable.

thorio commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 **/*.rs.bk
 Cargo.lock
+
+*.code-workspace
```
That's not really an option as I sometimes _do_ want to commit this file, and ignoring it for all repositories would cause it to be forgotten about.

I've found a workaround by placing it in a nested, ignored folder, so this can be removed.

thorio commented on src/figment.rs:
```diff
@@ -410,6 +522,15 @@ impl Figment {
     ///     assert!(not_a_dict.extract_inner::<bool>("cat").is_err());
     ///     assert!(not_a_dict.extract_inner::<bool>("dog").is_err());
     ///
+    ///     let cat_lives = root.focus("lives.0");
+    ///     assert_eq!(cat_lives.extract_inner::<u8>("cat").unwrap(), 9);
+    ///     let dog_lives = root.focus("lives.1");
+    ///     assert_eq!(dog_lives.extract_inner::<u8>("dog").unwrap(), 1);
```
I wanted to demonstrate that key => index => key was possible, though I suppose I can simplify it.

thorio replied:
The additional array is actually crucial in demonstrating that focus works on arrays, since `focus("cat")` and `focus("cat.0")` both don't work.

```rust
figment::Jail::expect_with(|jail| {
    jail.create_file("Config.toml", r#"
        cat = [1, 2, 3]
        dog = [4, 5, 6]
    "#)?;

    let root = Figment::from(Toml::file("Config.toml"));
    assert_eq!(root.extract_inner::<u8>("cat.0").unwrap(), 1);
    assert_eq!(root.extract_inner::<u8>("dog.2").unwrap(), 6);

    let cat = root.focus("cat");
    // This should work if `root.extract("cat.1")` were equal to
    // `root.focus("cat").extract("1")` but this property only holds true
    // for dictionaries.
    assert_eq!(cat.extract_inner::<u8>("1").unwrap(), 2);

    Ok(())
});
```

So focusing _through_ an array onto a nested dictionary (`lives.0`) works, but focusing directly _on_ an array (`lives`) or a  non-dict in an array (`cat.0`) does not.
I think this is a reasonable restriction, but should be properly documented.

Thoughts?

thorio commented on src/value/value.rs:
```diff
@@ -116,30 +116,46 @@ impl Value {
-    ///     }
+    ///         },
+    ///     }),
+    ///     "duck" => Value::from(vec![
+    ///         map!{
+    ///             "pasta" => 42usize,
+    ///         },
+    ///         map!{
+    ///             "pizza" => 229usize,
+    ///         },
+    ///     ]),
```
It would involve either the array being fairly deeply nested and/or having an array of mixed types. I instead opted to remove one value from each top-level key.

thorio commented on src/util.rs:
```diff
@@ -270,14 +270,25 @@ use crate::value::{Value, Dict};
-                dict.into()
-            }
-            Some(_) | None => value
+        let Some(key) = keys.next() else {
```
What do you mean? This is just the base case, it's not any more special than the previous implementation.

thorio committed:
Fix PR issues

Round 1.
Once reviewed, these changes should be merged into the appropriate
commits to produce a clean history.

thorio commented on src/coalesce.rs:
```diff
@@ -27,9 +29,10 @@ impl Coalescible for Value {
         use {Value::Dict as D, Value::Array as A, Order::*};
         match (self, other, o) {
-            (D(t, a), D(_, b), Join | Adjoin) | (D(_, a), D(t, b), Merge | Admerge) => D(t, a.coalesce(b, o)),
+            (D(t, a), D(_, b), Join | Adjoin | Zipjoin) | (D(_, a), D(t, b), Merge | Admerge | Zipmerge) => D(t, a.coalesce(b, o)),
```
I split the join and merge cases into separate lines, I think this is more readable than any wrapping option.

thorio commented:
I fixed all the issues with obvious solutions, I think the rest warrant further discussion.

The big benefit with the zip- strategies is the handling of "there is no value here" values, without it the Env provider would not work correctly for `MYARRAY.0=10`, `MYARRAY.1=20`. I picked `None` for this because it was the simplest option, as well as being fairly pointless in arrays (you'd just remove the entry and shorten the array instead of putting `None` there).

This also touches on #112, where the user wanted a way to specify "there is no value here", such that `None` would _not_ replace the previous `Some` value.

So we need to differentiate between these three cases when `merge`-ing:
- "I have an incoming value and I want to replace the existing value" => `Some(value)`
- "I do not have an incoming value and want to remove the existing value" => `None`
- "I do not have an incoming value and want to leave existing value untouched" => ???

`join`-ing works the same way, with "incoming value" and "existing value" swapped.

Thus I think we do need something like `Value::Void` that will always be replaced, regardless of strategy, such that
```
            [1, 2, _, 4] 
merged with [_, 8, 3, _]
becomes     [1, 8, 3, 4]        where _ is this "void" value.

lmmx commented:
For anyone who was waiting to use this I've merged this in the figment2 fork as a standalone release (0.11.2), please let me know if there was anything that was not considered ready and I can take a look but the reviews seem to have been mostly addressed here, give or take some remaining disagreement over the 'special cases'.

- https://github.com/lmmx/figment2/pull/6
